### PR TITLE
Fix issue with simulate_data

### PR DESCRIPTION
**Pull request by mdoucet**
The was a problem using simulate_data() with magnetic samples. Passing an array with a dR for each q value resulted in only the first dR value being used.

### CODE REVIEW

File: refl1d/probe.py [hunk 1]
```diff
@@ -1416,7 +1414,8 @@ def restore_data(self):
     restore_data.__doc__ = Probe.restore_data.__doc__
 
     def simulate_data(self, theory, noise=2):
-        if numpy.isscalar(noise):
+        noise = numpy.asarray(noise)
+        if len(noise.shape) < 2:
```
[the remark this answers is not in the document]
Comment (pkienzle) [reply refl1d/probe.py]: Further: need to explicitly test for None.  Otherwise, `np.asarray(None)` => `array(None, dtype=object)`, which returns false for `(noise is None)`.

File: refl1d/probe.py [hunk 2]
```diff
@@ -287,9 +287,7 @@ def simulate_data(self, theory, noise=None):
         if noise is None:
             pass
         else:
-            self.dR = numpy.asarray(noise)
-            if len(self.dR.shape) == 0:  # noise is a scalar
-                self.dR = 0.01 * noise * self.Ro
+            self.dR = 0.01 * numpy.asarray(noise) * self.Ro
```
[the remark this answers is not in the document]
Comment (mdoucet) [reply refl1d/probe.py]: Yes, your last point was why we wanted to be able to supply an array. It's to deal with the case where a simple dR/R doesn't work and we want to supply our own.

File: refl1d/probe.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mdoucet) [reply refl1d/probe.py]: Those are good points. There's not reason to convert to np arrays other than as a convenience to check the dimensions.

It's true that the different possibilities for the input dimensions create a little bit of a headache. I would almost vote for not dealing with four-tuples and use the same noise (array or scalar) for all cross-sections. That's likely the way it will be used. But check the extended condition I just added as a possible solution that allows all the scenarios. Note that we're not complaining if the dimensions are not compatible with the data... I didn't add it because it wasn't there before. We might want to add the extra check to verify that the length of dR is the length of Q.

**Comment by pkienzle** [comment]
I did a little linting.  Also, I switched the magnetic test to use the Magnetism property on slabs instead of MagneticSlab in the test.  Feel free to merge.

**Comment by pkienzle** [comment]
I twiddled the interface to simulate_data a little more.

I don't like that noise=None is not equivalent to skipping the noise parameter in the function call, but I don't know what we can do to fix it.  

I also tried to figure out a way to set the simulated data to the exact theory value while still specifying noise=dR/R%, but no luck.  In the end I made noise<=0. an error condition.  I changed your tests to set noise at 1e-11 instead.

I dumped it all in master along with a couple of other changes.  Hope it doesn't break anything.
